server/src/linter.ts

```diff
@@ -0,0 +1,199 @@
+export type ShellcheckComment = Readonly<{
+  file: string
+  line: number // 1-based
+  endLine: number // 1-based
+  column: number // 1-based
+  endColumn: number // 1-based
+  level: string // See mapShellcheckServerity
+  code: number
+  message: string
+
+  // The Fix data type appears to be defined here. We aren't making use of
+  // it yet but this might help down the road:
+  // https://github.com/koalaman/shellcheck/blob/4e703e5c61c6366bfd486d728bc624025e344e68/src/ShellCheck/Interface.hs#L271
+  // fix: any;
+}>
+
+export function assertShellcheckResult(val: any): asserts val is ShellcheckResult {
+  if (val !== null && typeof val !== 'object') {
+    throw new Error(`shellcheck: unexpected json output ${typeof val}`)
+  }
+
+  if (!Array.isArray(val.comments)) {
+    throw new Error(
+      `shellcheck: unexpected json output: expected 'comments' array ${typeof val.comments}`,
+    )
+  }
+
+  for (const idx in val.comments) {
+    const comment = val.comments[idx]
+    if (comment !== null && typeof comment != 'object') {
+      throw new Error(
+        `shellcheck: expected comment at index ${idx} to be object, found ${typeof comment}`,
+      )
+    }
+    if (typeof comment.file !== 'string')
+      throw new Error(
+        `shellcheck: expected comment file at index ${idx} to be string, found ${typeof comment.file}`,
+      )
+    if (typeof comment.level !== 'string')
+      throw new Error(
+        `shellcheck: expected comment level at index ${idx} to be string, found ${typeof comment.level}`,
+      )
+    if (typeof comment.message !== 'string')
+      throw new Error(
+        `shellcheck: expected comment message at index ${idx} to be string, found ${typeof comment.level}`,
+      )
+    if (typeof comment.line !== 'number')
+      throw new Error(
+        `shellcheck: expected comment line at index ${idx} to be number, found ${typeof comment.line}`,
+      )
+    if (typeof comment.endLine !== 'number')
+      throw new Error(
+        `shellcheck: expected comment endLine at index ${idx} to be number, found ${typeof comment.endLine}`,
+      )
+    if (typeof comment.column !== 'number')
+      throw new Error(
+        `shellcheck: expected comment column at index ${idx} to be number, found ${typeof comment.column}`,
+      )
+    if (typeof comment.endColumn !== 'number')
+      throw new Error(
+        `shellcheck: expected comment endColumn at index ${idx} to be number, found ${typeof comment.endColumn}`,
+      )
+    if (typeof comment.code !== 'number')
+      throw new Error(
+        `shellcheck: expected comment code at index ${idx} to be number, found ${typeof comment.code}`,
+      )
+  }
+}
+
+// Severity mappings:
+// https://github.com/koalaman/shellcheck/blob/364c33395e2f2d5500307f01989f70241c247d5a/src/ShellCheck/Formatter/Format.hs#L50
+// prettier-ignore
+const mapShellcheckServerity = (sev: string): LSP.DiagnosticSeverity =>
```

Comment:
This is a matter of taste but the super nested ternary is harder to read IMO than:

```typescript
const severityMapping: Record<string, undefined | LSP.DiagnosticSeverity> = {
    error: LSP.DiagnosticSeverity.Error,    
    warning: ...
}
   
const map = severityMapping[sev] || LSP.DiagnosticSeverity.Error

```